providers/dns/cpanel/cpanel.go

```diff
@@ -0,0 +1,346 @@
+
+	var found bool
+	var existingRecord shared.ZoneRecord
+	for _, record := range zoneInfo {
+		if contains(record.DataB64, valueB64) {
+			existingRecord = record
+			found = true
+			break
+		}
+	}
+
+	if !found {
+		return nil
+	}
+
+	var newData []string
+	for _, dataB64 := range existingRecord.DataB64 {
+		if dataB64 == valueB64 {
+			continue
+		}
+
+		data, errD := base64.StdEncoding.DecodeString(dataB64)
+		if errD != nil {
+			return fmt.Errorf("cpanel[mode=%s]: decode base64 record value: %w", d.config.Mode, errD)
+		}
+
+		newData = append(newData, string(data))
+	}
+
+	// Delete record.
+	if len(newData) == 0 {
+		_, err = d.client.DeleteRecord(ctx, serial, zone, existingRecord.LineIndex)
+		if err != nil {
+			return fmt.Errorf("cpanel[mode=%s]: delete record: %w", d.config.Mode, err)
+		}
+
+		return nil
+	}
+
+	// Remove one value.
+	record := shared.Record{
+		DName:      info.EffectiveFQDN,
+		TTL:        d.config.TTL,
+		RecordType: "TXT",
+		Data:       newData,
+		LineIndex:  existingRecord.LineIndex,
+	}
+
+	_, err = d.client.EditRecord(ctx, serial, zone, record)
+	if err != nil {
+		return fmt.Errorf("cpanel[mode=%s]: edit record: %w", d.config.Mode, err)
+	}
+
+	return nil
+}
+
+func getZoneSerial(zoneFqdn string, zoneInfo []shared.ZoneRecord) (uint32, error) {
+	nameB64 := base64.StdEncoding.EncodeToString([]byte(zoneFqdn))
+
+	for _, record := range zoneInfo {
+		if record.Type != "record" || record.RecordType != "SOA" || record.DNameB64 != nameB64 {
+			continue
+		}
+
+		// https://github.com/go-acme/lego/issues/1060#issuecomment-1925572386
+		// https://github.com/go-acme/lego/issues/1060#issuecomment-1925581832
+		data, err := base64.StdEncoding.DecodeString(record.DataB64[2])
+		if err != nil {
+			return 0, fmt.Errorf("decode serial DNameB64: %w", err)
+		}
+
+		var newSerial uint32
+		_, err = fmt.Sscan(string(data), &newSerial)
+		if err != nil {
+			return 0, fmt.Errorf("decode serial DNameB64, invalid serial value %q: %w", string(data), err)
+		}
+
+		return newSerial, nil
+	}
+
+	return 0, errors.New("zone serial not found")
+}
+
+func createClient(config *Config) (apiClient, error) {
+	switch strings.ToLower(config.Mode) {
+	case "cpanel":
+		client, err := cpanel.NewClient(config.BaseURL, config.Username, config.Token)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create cPanel API client: %w", err)
+		}
+
+		if config.HTTPClient != nil {
+			client.HTTPClient = config.HTTPClient
+		}
+
+		return client, nil
+
+	case "whm":
+		client, err := whm.NewClient(config.BaseURL, config.Username, config.Token)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create WHM API client: %w", err)
+		}
+
+		if config.HTTPClient != nil {
+			client.HTTPClient = config.HTTPClient
+		}
+
+		return client, nil
+
+	default:
+		return nil, fmt.Errorf("unsupported mode: %q", config.Mode)
+	}
+}
+
+func contains(values []string, value string) bool {
```

Comment:
There's [`slices.Contains`](https://pkg.go.dev/slices#Contains), which does the same. I'm not sure which Go version brought that it, so we might not use it yet.

Reply:
Looks like this is available since Go 1.21: https://pkg.go.dev/slices?tab=versions